clippy_lints/src/format_args.rs

```diff
@@ -559,6 +609,71 @@ impl<'tcx> FormatArgsExpr<'_, 'tcx> {
 
         false
     }
+
+    fn has_pointer_debug(&mut self, ty: Ty<'tcx>) -> bool {
+        let cx = self.cx;
+        let tcx = cx.tcx;
+        let typing_env = cx.typing_env();
+        let ty = tcx.normalize_erasing_regions(typing_env, ty);
+        if let Some(known) = self.has_pointer_format.get(&ty) {
+            return *known;
+        }
+        let mut open: Vec<Ty<'_>> = Vec::new();
+        open.push(ty);
+        while let Some(next_ty) = open.pop() {
+            if let Some(known) = self.has_pointer_format.get(&next_ty) {
+                if *known {
+                    return true;
+                }
+                continue;
+            }
+            match next_ty.kind() {
+                ty::RawPtr(..) | ty::FnPtr(..) | ty::FnDef(..) => {
+                    self.has_pointer_format.insert(ty, true);
+                    return true;
+                },
+                ty::Ref(_, t, _) | ty::Slice(t) | ty::Array(t, _) => {
+                    open.push(tcx.normalize_erasing_regions(typing_env, *t));
+                },
+                ty::Tuple(ts) => open.extend(ts.iter().map(|ty| tcx.normalize_erasing_regions(typing_env, ty))),
```

Comment:
Since the type is already normalized you don't need to normalize the contained types here. It's only ADT fields that normalization doesn't affect.